File: portal/docs/templates/backend/service.py
# -*- coding: utf-8 -*-
"""
TODO: 替换 <Resource> 为模型类名
TODO: 替换 <resource> 为资源名（下划线）
"""

from backend.app.<domain>.crud.<resource> import <Resource>CRUD

# TODO: 按实际字段调整（用于导出列头 & 导入列头映射）
EXPORT_FIELD_MAP = {
    'id': 'ID',
    'name': '名称',
    # TODO: 补充其他字段，如 'status': '状态'
    'created_at': '创建时间',
}

IMPORT_HEADER_MAP = {
    '名称': 'name',
    # TODO: 补充其他可导入字段
}


class <Resource>ServiceError(Exception):
    def __init__(self, message, status_code=400, payload=None):
        super().__init__(message)
        self.message = message
        self.status_code = status_code
        self.payload = payload or {}


class <Resource>Service:
    def __init__(self, db, model):
        self.db = db
        self.crud = <Resource>CRUD(db, model)

    # ── 列表 ──────────────────────────────────────────────────────────────────

    def list_items(self, page=1, per_page=20, search=''):
        query = self.crud.query_all()
        if search:
            query = query.filter(self.crud.<Resource>.name.ilike(f'%{search}%'))
        pagination = query.order_by(self.crud.<Resource>.id.desc()).paginate(
            page=page, per_page=per_page, error_out=False
        )
        return {
            'items': [item.to_dict() for item in pagination.items],
            'total': pagination.total,
        }

    # ── 新增 ──────────────────────────────────────────────────────────────────

    def create_item(self, data):
        if not data.get('name'):
            raise <Resource>ServiceError('名称不能为空', 400)
        # TODO: 补充唯一性校验（如需要）
        item = self.crud.<Resource>(
            name=data['name'],
            # TODO: 补充其他字段
        )
        try:
            self.crud.add(item)
            self.crud.commit()
            return item.to_dict(), 201
        except Exception as e:
            self.crud.rollback()
            raise <Resource>ServiceError(str(e), 500) from e

    # ── 编辑 ──────────────────────────────────────────────────────────────────

    def update_item(self, item, data):
        # TODO: 补充可更新字段
        if 'name' in data:
            item.name = data['name']
        try:
            self.crud.commit()
            return item.to_dict()
        except Exception as e:
            self.crud.rollback()
            raise <Resource>ServiceError(str(e), 500) from e

    # ── 删除 ──────────────────────────────────────────────────────────────────

    def delete_item(self, item):
        try:
            self.crud.delete(item)
            self.crud.commit()
            return {'message': '删除成功'}
        except Exception as e:
            self.crud.rollback()
            raise <Resource>ServiceError(str(e), 500) from e

    # ── 导出 ──────────────────────────────────────────────────────────────────

    def export_items(self, data):
        from backend.common.tabular import build_table_response, normalize_table_file_type
        file_type = normalize_table_file_type(data.get('file_type'), default='xlsx')
        fields = data.get('fields') or list(EXPORT_FIELD_MAP.keys())
        ids = data.get('ids')

        query = self.crud.query_all()
        if ids:
            query = query.filter(self.crud.<Resource>.id.in_(ids))
        items = query.order_by(self.crud.<Resource>.id.desc()).all()

        headers = [EXPORT_FIELD_MAP.get(f, f) for f in fields]
        rows = [[item.to_dict().get(f, '') for f in fields] for item in items]
        return build_table_response(headers, rows, '<resource>_export', file_type=file_type)

    # ── 下载导入模板 ──────────────────────────────────────────────────────────

    def download_template(self, file_type_raw='xlsx'):
        from backend.common.tabular import build_table_response, normalize_table_file_type
        file_type = normalize_table_file_type(file_type_raw, default='xlsx')
        headers = list(IMPORT_HEADER_MAP.keys())
        return build_table_response(headers, [], '<resource>_import_template', file_type=file_type)

    # ── 导入 ──────────────────────────────────────────────────────────────────

    def import_items(self, file_storage):
        from backend.common.tabular import read_table_file
        try:
            _, rows, _ = read_table_file(file_storage)
        except ValueError as e:
            raise <Resource>ServiceError(str(e), 400) from e

        created = 0
        error_rows = []
        for line, row in rows:
            try:
                name = row.get(IMPORT_HEADER_MAP.get('名称', ''), '').strip()
                if not name:
                    error_rows.append({'line': line, 'reason': '名称不能为空', 'row': row})
                    continue
                # TODO: 补充其他字段赋值
                item = self.crud.<Resource>(name=name)
                self.crud.add(item)
                created += 1
            except Exception as e:
                error_rows.append({'line': line, 'reason': str(e), 'row': row})

        if created:
            try:
                self.crud.commit()
            except Exception as e:
                self.crud.rollback()
                raise <Resource>ServiceError(str(e), 500) from e

        if error_rows:
            raise <Resource>ServiceError('部分数据导入失败', 400, {'error_rows': error_rows})
        return {'created': created, 'updated': 0}

File: apps/api/src/modules/admin/announcement/service.ts
/**
 * Announcement management service layer
 */

import { sql, type SQL } from 'drizzle-orm'
import { ServiceError } from '@/common/errors'
import { notFound } from '@/common/http'
import { pyInt, pyStr, pyTruthy } from '@/common/py'
import { buildTable, normalizeTableFileType, readTableFile, TableFileError, type UploadedFile } from '@/common/tabular'
import type { Db } from '@/db/client'
import { utcNow } from '@/db/schema/columns'
import { announcementToDict, type Announcement } from '@/db/schema'
import { bindText, omitNull, pyEq } from '@/common/sqla-bind'
import { AnnouncementRepository, type AnnouncementFilters, type AnnouncementUpdate } from './repository'
import {
  ANNOUNCE_TYPES,
  EXPORT_FIELD_MAP,
  IMPORT_HEADER_MAP,
  STATUSES,
  TEMPLATE_HEADERS,
  TEMPLATE_ROWS,
  idsForInClause,
  normalizeDbTimestamp,
  parsePublishAt,
  pickExportFields,
  stripOrEmpty,
  type PyDateTime,
} from './schema'

type Data = Record<string, unknown>

/** Integer conversion failures are treated as uncaught errors → global 500 */
function pyIntOr500(value: unknown): number {
  try {
    return pyInt(value)
  } catch (err) {
    throw new ServiceError(err instanceof Error ? err.message : String(err), 500)
  }
}

/** Value to write for publish_at: parsed result / current UTC time / None */
type PublishAtValue = PyDateTime | 'now' | null

function publishAtSql(value: PublishAtValue): SQL | string | null {
  if (value === null) return null
  if (value === 'now') return utcNow()
  // aware datetimes are sent as `'...'::timestamptz` and converted to the session time zone when stored in a timestamp column
  return value.aware ? sql`${value.text}::timestamptz` : value.text
}

/** Whether the publish time is unchanged: naive values compare by value; aware values never equal the naive DB value */
function publishAtEquals(value: PublishAtValue, current: string | null): boolean {
  if (value === null) return current === null
  if (value === 'now' || value.aware) return false
  return value.text === normalizeDbTimestamp(current)
}

/** SQLSTATE → error class name, used as the prefix of import row error reasons (keeps the existing error message format) */
const PG_ERROR_CLASS: Record<string, string> = {
  '22001': 'StringDataRightTruncation',
  '22003': 'NumericValueOutOfRange',
  '22P02': 'InvalidTextRepresentation',
  '23502': 'NotNullViolation',
}

function importRowReason(err: unknown): string {
  // Drizzle wraps the pg error in `cause`
  const e = ((err as { cause?: unknown }).cause ?? err) as { code?: string; message?: string }
  const cls = e.code && PG_ERROR_CLASS[e.code]
  const message = e.message ?? String(err)
  return cls ? `(psycopg2.errors.${cls}) ${message}` : message
}

export class AnnouncementService {
  private readonly repo: AnnouncementRepository

  constructor(private readonly db: Db) {
    this.repo = new AnnouncementRepository(db)
  }

  async getOr404(id: number): Promise<Announcement> {
    const item = await this.repo.get(id)
    if (!item) throw notFound()
    return item
  }

  async listItems(page: number, perPage: number, filters: AnnouncementFilters) {
    const { total, rows } = await this.repo.listPage(page, perPage, filters)
    return { items: rows.map(announcementToDict), total }
  }

  async createItem(data: Data) {
    const title = stripOrEmpty(data.title)
    if (!title) throw new ServiceError('标题不能为空', 400)

    const publishAt = pyTruthy(data.publish_at) ? parsePublishAt(data.publish_at) : null

    const content = bindText(pyTruthy(data.content) ? data.content : '')
    const announceType = bindText(pyTruthy(data.announce_type) ? data.announce_type : 'system')
    const status = bindText(pyTruthy(data.status) ? data.status : 'draft')
    const isTop = pyTruthy('is_top' in data ? data.is_top : false)
    const sortOrder = pyIntOr500(pyTruthy(data.sort_order) ? data.sort_order : 0)

    try {
      const created = await this.repo.insert({
        title,
        content: omitNull(content),
        announce_type: omitNull(announceType),
        status: omitNull(status),
        is_top: isTop,
        sort_order: sortOrder,
        publish_at: omitNull(publishAtSql(publishAt)) as string | undefined,
      })
      return announcementToDict(created)
    } catch (err) {
      throw new ServiceError(err instanceof Error ? err.message : String(err), 500)
    }
  }

  /** Compute final values in a fixed field order and write back only columns that differ from the current value (`==` semantics); no UPDATE when nothing changed */
  private async applyChanges(item: Announcement, assigned: Record<string, unknown>, publishAt?: PublishAtValue) {
    const changes: AnnouncementUpdate = {}
    for (const [field, value] of Object.entries(assigned)) {
      if (!pyEq(value, (item as Record<string, unknown>)[field])) {
        ;(changes as Record<string, unknown>)[field] = value
      }
    }
    if (publishAt !== undefined && !publishAtEquals(publishAt, item.publish_at)) {
      changes.publish_at = publishAtSql(publishAt) as string | null
    }
    if (Object.keys(changes).length === 0) return announcementToDict(item)

    // Bind by column type before writing (PG assignment-cast semantics)
    for (const field of ['content', 'announce_type', 'status'] as const) {
      if (field in changes) changes[field] = bindText(changes[field]) as never
    }
    try {
      return announcementToDict(await this.repo.update(item.id, changes))
    } catch (err) {
      throw new ServiceError(err instanceof Error ? err.message : String(err), 500)
    }
  }

  async updateItem(item: Announcement, data: Data) {
    const assigned: Record<string, unknown> = {}
    let publishAt: PublishAtValue | undefined

    if ('title' in data) {
      const title = stripOrEmpty(data.title)
      if (!title) throw new ServiceError('标题不能为空', 400)
      assigned.title = title
    }
    if ('content' in data) assigned.content = pyTruthy(data.content) ? data.content : ''
    if ('announce_type' in data) assigned.announce_type = data.announce_type
    if ('status' in data) {
      assigned.status = data.status
      if (data.status === 'published' && item.publish_at === null) publishAt = 'now'
    }
    if ('is_top' in data) assigned.is_top = pyTruthy(data.is_top)
    if ('sort_order' in data) assigned.sort_order = pyIntOr500(pyTruthy(data.sort_order) ? data.sort_order : 0)
    if ('publish_at' in data) {
      if (pyTruthy(data.publish_at)) {
        const parsed = parsePublishAt(data.publish_at)
        if (parsed) publishAt = parsed
      } else {
        publishAt = null
      }
    }
    return this.applyChanges(item, assigned, publishAt)
  }

  async deleteItem(item: Announcement) {
    try {
      await this.repo.delete(item.id)
    } catch (err) {
      throw new ServiceError(err instanceof Error ? err.message : String(err), 500)
    }
    return { message: '删除成功' }
  }

  async publishItem(item: Announcement) {
    return this.applyChanges(item, { status: 'published' }, item.publish_at === null ? 'now' : undefined)
  }

  async unpublishItem(item: Announcement) {
    return this.applyChanges(item, { status: 'draft' })
  }

  exportFields() {
    return Object.entries(EXPORT_FIELD_MAP).map(([value, [label]]) => ({ label, value }))
  }

  async exportItems(data: Data) {
    const fieldsRaw = pyTruthy(data.fields) ? data.fields : []
    const fileType = normalizeTableFileType(data.file_type, 'xlsx')
    const ids = pyTruthy(data.ids) ? data.ids : []
    const exportMode = pyStr(pyTruthy(data.export_mode) ? data.export_mode : 'all').trim()

    let validFields = pickExportFields(fieldsRaw)
    if (validFields.length === 0) validFields = Object.keys(EXPORT_FIELD_MAP)

    const items =
      exportMode === 'selected' && pyTruthy(ids)
        ? await this.repo.listByIdsOrdered(idsForInClause(ids))
        : await this.repo.listAllForExport()

    const headers = validFields.map((f) => EXPORT_FIELD_MAP[f]![0])
    const rows = items.map((item) => validFields.map((f) => EXPORT_FIELD_MAP[f]![1](item)))
    return buildTable(headers, rows, 'announcements_export', fileType)
  }

  async downloadTemplate(fileTypeRaw: unknown) {
    return buildTable(TEMPLATE_HEADERS, TEMPLATE_ROWS, 'announcements_import_template', normalizeTableFileType(fileTypeRaw, 'xlsx'))
  }

  /** Commit row by row: successful rows are kept, failed rows go to error_rows (no overall rollback) */
  async importItems(file: UploadedFile | null) {
    if (!file) throw new ServiceError('请上传导入文件', 400)
    let table
    try {
      table = await readTableFile(file)
    } catch (err) {
      if (err instanceof TableFileError) throw new ServiceError(err.message, 400)
      throw err
    }
    if (table.fieldnames.length === 0) throw new ServiceError('文件为空或格式错误', 400)

    // Note: keys are the stripped headers, but lookups check the raw headers (a header with surrounding whitespace is not mapped; preserves existing behavior)
    const colMap = new Map<string, string>()
    for (const col of table.fieldnames) {
      const key = col.trim()
      if (Object.hasOwn(IMPORT_HEADER_MAP, key)) colMap.set(key, IMPORT_HEADER_MAP[key]!)
    }

    let created = 0
    const errorRows: { line: number; reason: string; row: Record<string, string> }[] = []
    for (const [line, row] of table.rows) {
      const mapped: Record<string, string> = {}
      for (const [k, v] of Object.entries(row)) {
        const field = colMap.get(k)
        if (field) mapped[field] = v
      }
      const title = (mapped.title || '').trim()
      if (!title) {
        errorRows.push({ line, reason: '标题不能为空', row })
        continue
      }
      const isTop = ['是', 'true', 'True', '1'].includes((mapped.is_top || '').trim())
      let sortOrder: number
      try {
        sortOrder = pyInt(mapped.sort_order || 0)
      } catch {
        sortOrder = 0
      }
      let announceType = (mapped.announce_type || 'system').trim()
      if (!ANNOUNCE_TYPES.includes(announceType)) announceType = 'system'
      let status = (mapped.status || 'draft').trim()
      if (!STATUSES.includes(status)) status = 'draft'
      try {
        await this.repo.insert({
          title,
          content: mapped.content || '',
          announce_type: announceType,
          status,
          is_top: isTop,
          sort_order: sortOrder,
        })
        created += 1
      } catch (err) {
        errorRows.push({ line, reason: importRowReason(err), row })
      }
    }

    if (errorRows.length > 0) return { created, updated: 0, error_rows: errorRows }
    return { created, updated: 0 }
  }
}
